cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -66,12 +77,47 @@ public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
 
       //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
       return writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
-    } finally {
-      try {
-        repositoryManager.close();
-      } catch (IOException e) {
-        LOG.warn("Failed to close the RepositoryManager, there may be leftover files", e);
+    }
+  }
+
+  @Override
+  public PullAppResponse pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws PullFailureException,
+    NotFoundException, AuthenticationConfigException, IllegalArgumentException {
+    String applicationName = appRef.getApplication();
+    String configFileName = generateConfigFileName(applicationName);
+    LOG.info("Cloning remote to pull application {}", applicationName);
+    String contents;
+    String fileHash;
+    Path appRelativePath;
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(appRef.getNamespaceId(), repoConfig)) {
+      appRelativePath = repositoryManager.getFileRelativePath(configFileName);
+      String commitID = repositoryManager.cloneRemote();
+      Path filePathToRead;
+      filePathToRead = validateAppConfigRelativePath(repositoryManager, appRelativePath);
+      if (!Files.exists(filePathToRead)) {
+        throw new NotFoundException(String.format("App with name %s not found in git repository", applicationName));
       }
+      contents = new String(Files.readAllBytes(filePathToRead), StandardCharsets.UTF_8);
```

Comment:
Maybe we could try 
```
AppRequest<?> appRequest = DECODE_GSON.fromJson(Files.newBufferedReader(filePathToRead, StandardCharsets.UTF_8) , AppRequest.class);
```
And add the `JsonSyntaxException` handling in the following catch block
So that we could save the step of reading the file to a string.

Reply:
Is it really important to close repo manager before decoding? Of what's the reason of separating the try clauses?